modules/videoio/src/cap_libcamera.cpp

```diff
@@ -0,0 +1,1100 @@
+        controls_.set(controls::Sharpness, options_->sharpness);
+
+    if (camera_->start(&controls_))
+        throw std::runtime_error("failed to start camera");
+    controls_.clear();
+    camera_started_ = true;
+    last_timestamp_ = 0;
+
+    camera_->requestCompleted.connect(this, &LibcameraCapture::requestComplete);
+
+    for (std::unique_ptr<Request> &request : requests_)
+    {
+        if (camera_->queueRequest(request.get()) < 0)
+            throw std::runtime_error("Failed to queue request");
+    }
+
+    if (options_->verbose)
+        std::cerr << "Camera started!" << std::endl;
+}
+
+void LibcameraCapture::StopCamera()
+{
+    {
+        // We don't want QueueRequest to run asynchronously while we stop the camera.
+        std::lock_guard<std::mutex> lock(camera_stop_mutex_);
+        if (camera_started_)
+        {
+            std::cerr << "Camera tries to stop!!" << std::endl;
+            if (camera_->stop())
+                throw std::runtime_error("failed to stop camera");
+
+            camera_started_ = false;
+        }
+        // camera_->requestCompleted.disconnect(this, &LibcameraApp::requestComplete);
+        // if (!camera_->requestCompleted.disconnect(this, &LibcameraApp::requestComplete)) {
+        //     throw std::runtime_error("failed to disconnect camera callbacks");
+        // }
+    }
+
+    if (camera_)
+        camera_->requestCompleted.disconnect(this, &LibcameraCapture::requestComplete);
+
+    // An application might be holding a CompletedRequest, so queueRequest will get
+    // called to delete it later, but we need to know not to try and re-queue it.
+    completed_requests_.clear();
+
+    msg_queue_.Clear();
+
+    while (!free_requests_.empty())
+        free_requests_.pop();
+
+    requests_.clear();
+
+    controls_.clear(); // no need for mutex here
+
+    if (options_->verbose && !options_->help)
+        std::cerr << "Camera stopped!" << std::endl;
+}
+
+void LibcameraCapture::ApplyRoiSettings()
+{
+    if (!controls_.get(controls::ScalerCrop) && options_->roi_width != 0 && options_->roi_height != 0)
+    {
+        Rectangle sensor_area = *camera_->properties().get(properties::ScalerCropMaximum);
+        int x = options_->roi_x * sensor_area.width;
+        int y = options_->roi_y * sensor_area.height;
+        int w = options_->roi_width * sensor_area.width;
+        int h = options_->roi_height * sensor_area.height;
+        Rectangle crop(x, y, w, h);
+        crop.translateBy(sensor_area.topLeft());
+        if (options_->verbose)
+            std::cerr << "Using crop " << crop.toString() << std::endl;
+        controls_.set(controls::ScalerCrop, crop);
+    }
+}
+
+LibcameraCapture::Msg LibcameraCapture::Wait()
+{
+    return msg_queue_.Wait();
+}
+
+void LibcameraCapture::queueRequest(CompletedRequest *completed_request)
+{
+    BufferMap buffers(std::move(completed_request->buffers));
+
+    Request *request = completed_request->request;
+    assert(request);
```

Comment:
Please use CV_Assert. Plain C assert does not work in release by default. Use CV_Dbg_Assert, if it's debug only check.